snow/engine/snowman/straggler_detect.go

```diff
@@ -0,0 +1,331 @@
+				return false
+			}
+			processingWeight = newProcessingWeight
+		}
+	}
+
+	sd.log.Trace("Counted total weight that accepted blocks we're still processing", zap.Uint64("weight", processingWeight))
+
+	ratio := float64(processingWeight) / float64(totalValidatorWeight)
+
+	if ratio > stakeThresholdForStragglerSuspicion {
+		sd.log.Trace("We are straggling behind", zap.Float64("ratio", ratio))
+		return true
+	}
+
+	sd.log.Trace("Nodes ahead of us:", zap.Float64("ratio", ratio))
+
+	return false
+}
+
+func (sd *stragglerDetector) validateNetInfo(netInfo netInfo) bool {
+	if netInfo.connStakePercent < sd.minConfirmationThreshold {
+		// We don't know for sure whether we're behind or not.
+		// Even if we're behind, it's pointless to act before we have established
+		// connectivity with enough validators.
+		sd.log.Verbo("not enough connected stake to determine network info", zap.Float64("ratio", netInfo.connStakePercent))
+		return false
+	}
+
+	if netInfo.totalValidatorWeight == 0 {
+		sd.log.Trace("Connected to zero weight")
+		return false
+	}
+
+	totalKnownLastBlockStakePercent := float64(netInfo.totalWeightWeKnowItsLastAcceptedBlock) / float64(netInfo.totalValidatorWeight)
+	stakeAheadOfUs := float64(netInfo.totalPendingStake) / float64(netInfo.totalValidatorWeight)
+
+	// Ensure we have collected last accepted blocks for at least 80% (or so) stake of the total weight we are connected to.
+	if totalKnownLastBlockStakePercent < minimumStakeThresholdRequiredForNetworkInfo {
+		sd.log.Trace("Not collected enough information about last accepted blocks for the validators we are connected to",
+			zap.Float64("ratio", totalKnownLastBlockStakePercent))
+		return false
+	}
+
+	if stakeAheadOfUs < knownStakeThresholdRequiredForAnalysis {
+		sd.log.Trace("Most stake we're connected to has the same height as we do",
+			zap.Float64("ratio", stakeAheadOfUs))
+		return false
+	}
+
+	return true
+}
+
+func (sd *stragglerDetector) getNetworkSnapshot() (snapshot, bool) {
+	ourLastAcceptedBlock := sd.lastAccepted()
+
+	netInfo, err := sd.getNetworkInfo(ourLastAcceptedBlock)
+	if err != nil {
+		return snapshot{}, false
+	}
+
+	if !sd.validateNetInfo(netInfo) {
+		return snapshot{}, false
+	}
+
+	snap := snapshot{
+		nodeWeightsToBlocks:  netInfo.nodeWeightToLastAccepted,
+		totalValidatorWeight: netInfo.totalValidatorWeight,
+	}
+
+	if sd.haveWeFailedCatchingUp(snap) {
+		return snap, true
+	}
+
+	return snapshot{}, false
+}
+
+type netInfo struct {
+	connStakePercent                      float64
+	totalPendingStake                     uint64
+	totalValidatorWeight                  uint64
+	totalWeightWeKnowItsLastAcceptedBlock uint64
+	nodeWeightToLastAccepted              nodeWeightsToBlocks
+}
+
+func (sd *stragglerDetector) getNetworkInfo(ourLastAcceptedBlock ids.ID) (netInfo, error) {
+	var res netInfo
+
+	res.connStakePercent = sd.connectedPercent()
+
+	validators := nodeWeights(sd.connectedValidators().List())
```

Comment:
This seems racy with connectedPercent.

Reply:
it does not matter if the validators change after the connectedPercent was read.

We only use the connected stake to know if to abort early or continue the assessment. 

Let's look at two possibilities: 

1. The validators have more stake than connected percent, and we abort early in `validateNetInfo` because connected percent was below the threshold: Then, we will simply not detect whether we are behind in this probe. This scheduling is equivalent to a scheduling where we read the connected percent and the validators in closer temporal proximity.
2. We read a connected percent above the threshold, but afterwards, we read validators such that the connected percent is below the threshold: This is equivalent to the logic ignoring the connected percent altogether, and just relying on the data from `connectedValidators()`.

Initially, in the first version of this PR, the connected percent was the first thing we compute, and we wouldn't continue computing anything if it was too low. 

However, after the first code review, I changed the computation to first collect data, and then validate it.

I can remove the connected percent, as it might make the implementation simpler.

Reply:
I'd not oppose to have a simpler implementation if you say this is not ultra necessary.

Reply:
I removed it in the follow-up commit